pkg/cache/sql/informer/indexer.go

```diff
@@ -0,0 +1,258 @@
+package informer
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"reflect"
+	"strings"
+	"sync"
+
+	"github.com/pkg/errors"
+	"github.com/rancher/lasso/pkg/cache/sql/db"
+	"github.com/rancher/lasso/pkg/cache/sql/db/transaction"
+	"k8s.io/client-go/tools/cache"
+)
+
+const (
+	selectQueryFmt = `
+			SELECT object FROM "%s"
+				WHERE key IN (
+					SELECT key FROM "%s_indices"
+						WHERE name = ? AND value IN (?%s)
+				)
+		`
+	createTablefmt = `CREATE TABLE "%s_indices" (
+			name VARCHAR NOT NULL,
+			value VARCHAR NOT NULL,
+			key VARCHAR NOT NULL REFERENCES "%s"(key) ON DELETE CASCADE,
+			PRIMARY KEY (name, value, key)
+        )`
+	createIndexFmt = `CREATE INDEX "%s_indices_index" ON "%s_indices"(name, value)`
+
+	deleteIndicesFmt = `DELETE FROM "%s_indices" WHERE key = ?`
+	addIndexFmt      = `INSERT INTO "%s_indices"(name, value, key) VALUES (?, ?, ?)`
+	listByIndexFmt   = `SELECT object FROM "%s"
+			WHERE key IN (
+			    SELECT key FROM "%s_indices"
+			    	WHERE name = ? AND value = ?
+			)`
+	listKeyByIndexFmt  = `SELECT DISTINCT key FROM "%s_indices" WHERE name = ? AND value = ?`
+	listIndexValuesFmt = `SELECT DISTINCT value FROM "%s_indices" WHERE name = ?`
+)
+
+// Indexer is a SQLite-backed cache.Indexer which builds upon Store adding an index table
+type Indexer struct {
+	Store
+	indexers            cache.Indexers
+	indexersLock        sync.RWMutex
+	deleteIndicesStmt   *sql.Stmt
+	addIndexStmt        *sql.Stmt
+	listByIndexStmt     *sql.Stmt
+	listKeysByIndexStmt *sql.Stmt
+	listIndexValuesStmt *sql.Stmt
+}
+
+var _ cache.Indexer = (*Indexer)(nil)
+
+type Store interface {
+	DBClient
+	cache.Store
+
+	GetByKey(key string) (item any, exists bool, err error)
+	GetName() string
+	RegisterAfterUpsert(f func(key string, obj any, tx db.TXClient) error)
+	RegisterAfterDelete(f func(key string, tx db.TXClient) error)
+	GetShouldEncrypt() bool
+	GetType() reflect.Type
+}
+
+type DBClient interface {
+	Begin() (db.TXClient, error)
+	QueryForRows(ctx context.Context, stmt transaction.Stmt, params ...any) (*sql.Rows, error)
+	ReadObjects(rows db.Rows, typ reflect.Type, shouldDecrypt bool) ([]any, error)
+	ReadStrings(rows db.Rows) ([]string, error)
+	Prepare(stmt string) *sql.Stmt
+	CloseStmt(stmt db.Closable) error
+}
+
+// Test that Indexer implements cache.Indexervar _ cache.Indexer = (*Indexer)(nil)
+
+// NewIndexer returns a cache.Indexer backed by SQLite for objects of the given example type
+func NewIndexer(indexers cache.Indexers, s Store) (*Indexer, error) {
+	// sanity checks first
```

Comment:
`Store` is an interface, the use of `s.GetName()` below assumes that it depends on a concrete implementation which already sanitised its contents for SQL injection, which may not always hold true. This results in a few places using `s.GetName()` in this file being vulnerable to SQL injection.

Reply:
Well there is only one Store implementation, and it actually belongs to this library and package, and it does sanitize the name. The reason for having an interface here is mocking.

Are we really targeting a threat model in which a _programmer_ uses the library in an incorrect way, as opposed to a _user_ trying to inject SQL commands?

If we are, how important is this from your perspective?

Reply:
> Are we really targeting a threat model in which a programmer uses the library in an incorrect way, as opposed to a user trying to inject SQL commands?

I don't think those two are mutually exclusing. A programmer incorrectly using the library could lead to SQL injection down the line.

> The reason for having an interface here is mocking.

I'm still getting used to mocking for testing so sorry if that's a stupid question but do we need mocks here considering we're using sqlite as the only implementation and that it supports running in memory via `:memory:`?

If so and if we want to avoid programmer errors that lead to SQL injection, could we have this function be internal only and perhaps a less generic `NewIndexer` function for library use that only accepts the sqlite store? This way we can make sure that whoever is using lasso in rancher/k3s/rke2/wtv can only pass in a "safe" implementation.

Reply:
but we are similarly relying on indexed fields https://github.com/rancher/lasso/pull/65/files#diff-910681530ceac7cfe911fb307754542734adee41c8d009d795aabe0d9e3803b2R25 for validation which could also be configured to inject, and a developer could misuse partitions. A developer could expose any factory in lasso and that would also be unsafe. I just think lasso is covered in comparable cases where if you just used it as is it would be insecure to hand to a user. It’s an admin level tool.

I don’t think it hurts to move santize() to the db package, export it, and use it liberally where we perform sql queries. I went ahead and did that. I just don’t think the original logic tracks for me.

@tomleb I think it's better not to incur the overhead of running sqlite as the job of the unit tests are to test the functionality contained in the method. The store is also tested and sqlite is tested by the maintainers. We then have integration tests which test them working together. It also could affect running the tests in parallel.

Reply:
-> [epic](https://github.com/rancher/rancher/issues/45635)